### DISCUSSION TITLE
Failed to compile libpag on Linux

### DISCUSSION THREAD

**Discussion by sangenan** [Tencent/libpag#123] [Q&A] [answered]
### **environment dependency** 
gcc version 7.5.0
re2c 1.3
ninja 1.10.2
node v17.3.1

Execute in the root directory  ./build_linux.sh  and compile failed
`
In file included from /libpag-3.2.7.40/src/base/ByteData.cpp:19:0:
/libpag-3.2.7.40/tgfx/include/core/Stream.h:35:15: error: ‘unique_ptr’ in namespace ‘std’ does not name a template type
   static std::unique_ptr<Stream> MakeFromFile(const std::string& filePath);
               ^~~~~~~~~~
/libpag-3.2.7.40/src/base/ByteData.cpp: In static member function ‘static std::unique_ptr<pag::ByteData> pag::ByteData::FromPath(const string&)’:
/libpag-3.2.7.40/src/base/ByteData.cpp:24:25: error: ‘MakeFromFile’ is not a member of ‘pag::Stream’
   auto stream = Stream::MakeFromFile(filePath);
                         ^~~~~~~~~~~~
make[3]: *** [CMakeFiles/pag.dir/build.make:93: CMakeFiles/pag.dir/src/base/ByteData.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/Makefile2:76: CMakeFiles/pag.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/pag.dir/rule] Error 2
make: *** [Makefile:118: pag] Error 2
`

**Comment by kevingpqi123** [comment, marked as answer]
你再试下main主干最新的版本

**Reply by sangenan** [reply]
在Linux上拉取了main的主干最新分支，可以顺利编译，但linux下demo无法make成功
`
/usr/bin/ld:/home/Desktoplibpag/linux/vendor/swiftshader/linux/x64/libEGL.so: file format not recognized; treating as linker script
`
发现.so文件格式变成了ASCII text

**Reply by kevingpqi123** [reply]
你可以到这个路径下载依赖动态库：https://github.com/libpag/swiftshader/releases
剩余的就是动态库的配置了

**Reply by sangenan** [reply]
Thanks♪(･ω･)ﾉ

**Reply by sangenan** [reply]
@kevingpqi123 demo还是无法编译通过提示如下：
`
[100%] Linking CXX executable pag-linux
/usr/bin/ld: ../vendor/pag/linux/x64/libpag.a(ithread.c.o): undefined reference to symbol 'sem_destroy@@GLIBC_2.2.5'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libpthread.so: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/pag-linux.dir/build.make:88: pag-linux] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/pag-linux.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
`
